// impound/dist/index.mjs
import { init, parse } from "es-module-lexer";
import { isAbsolute, join, relative } from "pathe";
import { createUnplugin } from "unplugin";
import { createFilter } from "unplugin-utils";

//#region src/index.ts
const PROXY_ID = "\0impound:proxy";
const PROXY_CODE = `
function createMock(name, overrides = {}) {
  const proxyFn = function () {};
  proxyFn.prototype.name = name;
  const props = {};
  const proxy = new Proxy(proxyFn, {
    get(_target, prop) {
      if (prop === "caller") return null;
      if (prop === "__createMock__") return createMock;
      if (prop === "__mock__") return true;
      if (prop in overrides) return overrides[prop];
      if (prop === "then") return (fn) => Promise.resolve(fn());
      if (prop === "catch") return (_fn) => Promise.resolve();
      if (prop === "finally") return (fn) => Promise.resolve(fn());
      return (props[prop] = props[prop] || createMock(\`\${name}.\${prop.toString()}\`));
    },
    apply(_target, _this, _args) { return createMock(\`\${name}()\`); },
    construct(_target, _args, _newT) { return createMock(\`[\${name}]\`); },
    enumerate() { return []; },
  });
  return proxy;
}
export default createMock("mock");
`.trim();
const RELATIVE_IMPORT_RE = /^\.\.?\//;
/** Convert a byte offset in source code to a 1-indexed line and 0-indexed column. */
function offsetToLineColumn(code, offset) {
	let line = 1;
	let lastNewline = -1;
	for (let i = 0; i < offset && i < code.length; i++) if (code[i] === "\n") {
		line++;
		lastNewline = i;
	}
	return {
		line,
		column: offset - lastNewline - 1
	};
}
/** Generate a code snippet with context lines, a `>` marker, and a `^` caret. */
function generateSnippet(code, line, column, context = 2) {
	const lines = code.split("\n");
	const start = Math.max(0, line - 1 - context);
	const end = Math.min(lines.length, line + context);
	const gutterWidth = String(end).length;
	const result = [];
	for (let i = start; i < end; i++) {
		const lineNum = i + 1;
		const gutter = String(lineNum).padStart(gutterWidth);
		const marker = lineNum === line ? ">" : " ";
		result.push(`${marker} ${gutter} | ${lines[i]}`);
		if (lineNum === line) result.push(`  ${" ".repeat(gutterWidth)} | ${" ".repeat(column)}^`);
	}
	return result.join("\n");
}
/** Build an import trace from entry to the importer via BFS backwards through the graph. */
function buildTrace(importer, moduleGraph, resolvedImports, entries, maxDepth) {
	const visited = /* @__PURE__ */ new Set();
	const queue = [[importer, [importer]]];
	visited.add(importer);
	let bestPath = [importer];
	while (queue.length > 0) {
		const [current, path] = queue.shift();
		if (path.length > maxDepth) continue;
		if (entries.has(current)) {
			bestPath = path;
			break;
		}
		for (const [moduleId] of moduleGraph) {
			if (visited.has(moduleId)) continue;
			const resolvedForModule = resolvedImports.get(moduleId);
			if (resolvedForModule) {
				for (const [, resolvedId] of resolvedForModule) if (resolvedId === current) {
					visited.add(moduleId);
					const newPath = [...path, moduleId];
					if (entries.has(moduleId)) {
						bestPath = newPath;
						queue.length = 0;
						break;
					}
					queue.push([moduleId, newPath]);
					break;
				}
			}
		}
	}
	bestPath.reverse();
	const trace = [];
	for (let i = 0; i < bestPath.length; i++) {
		const file = bestPath[i];
		const step = { file };
		if (i === 0 && entries.has(file)) {}
		if (i < bestPath.length - 1) {
			const nextFile = bestPath[i + 1];
			const resolvedForFile = resolvedImports.get(file);
			const graphEntry = moduleGraph.get(file);
			if (resolvedForFile && graphEntry) {
				for (const [specifier, resolvedId] of resolvedForFile) if (resolvedId === nextFile) {
					step.import = specifier;
					const loc = graphEntry.imports.get(specifier);
					if (loc) {
						step.line = loc.line;
						step.column = loc.column;
					}
					break;
				}
			}
		}
		trace.push(step);
	}
	return trace;
}
function formatTrace(trace, cwd) {
	return trace.map((step, i) => {
		const file = cwd && isAbsolute(step.file) ? relative(cwd, step.file) : step.file;
		const loc = step.line != null ? `:${step.line}:${step.column ?? 0}` : "";
		const entry = i === 0 ? " (entry)" : "";
		const imp = step.import ? ` (import "${step.import}")` : "";
		return `  ${i + 1}. ${file}${loc}${entry}${imp}`;
	}).join("\n");
}
const ImpoundPlugin = createUnplugin((globalOptions) => {
	const matchers = "matchers" in globalOptions ? globalOptions.matchers : [globalOptions];
	const traceEnabled = globalOptions.trace === true;
	const maxTraceDepth = globalOptions.maxTraceDepth ?? 20;
	const moduleGraph = /* @__PURE__ */ new Map();
	const resolvedImports = /* @__PURE__ */ new Map();
	const entries = /* @__PURE__ */ new Set();
	const pendingViolations = [];
	const plugins = matchers.map((options) => {
		const filter = createFilter(options.include, options.exclude, { resolve: globalOptions.cwd });
		const excludeFilter = options.excludeFiles?.length ? createFilter(options.excludeFiles, void 0, { resolve: globalOptions.cwd }) : void 0;
		const warnedMessages = options.warn !== "always" ? /* @__PURE__ */ new Set() : void 0;
		return {
			name: "impound",
			enforce: "pre",
			load(id) {
				if (id === PROXY_ID) return PROXY_CODE;
			},
			resolveId(id, importer, resolveOptions) {
				if (id === PROXY_ID) return id;
				if (!importer) {
					if (traceEnabled && resolveOptions?.isEntry) entries.add(id);
					return;
				}
				if (!filter(importer)) return;
				const rawId = id;
				if (RELATIVE_IMPORT_RE.test(id)) id = join(importer, "..", id);
				if (excludeFilter?.(id)) return;
				if (isAbsolute(id) && globalOptions.cwd) id = relative(globalOptions.cwd, id);
				if (traceEnabled) {
					let importerResolved = resolvedImports.get(importer);
					if (!importerResolved) {
						importerResolved = /* @__PURE__ */ new Map();
						resolvedImports.set(importer, importerResolved);
					}
					importerResolved.set(rawId, id);
				}
				let matched = false;
				const relativeImporter = isAbsolute(importer) && globalOptions.cwd ? relative(globalOptions.cwd, importer) : importer;
				for (const [pattern, warning, suggestions] of options.patterns) {
					const usesImport = pattern instanceof RegExp ? pattern.test(id) : typeof pattern === "string" ? pattern === id : pattern(id, relativeImporter);
					if (usesImport) {
						const baseMessage = `${typeof usesImport === "string" ? usesImport : warning || "Invalid import"} [importing \`${id}\` from \`${relativeImporter}\`]`;
						if (traceEnabled) pendingViolations.push({
							id,
							rawId,
							importer,
							relativeImporter,
							message: baseMessage,
							suggestions,
							options
						});
						else {
							let message = baseMessage;
							if (suggestions?.length) message += `\n\nSuggestions:\n${suggestions.map((s) => `  - ${s}`).join("\n")}`;
							if (options.onViolation?.({
								id,
								importer: relativeImporter,
								message
							}) === false) continue;
							if (!warnedMessages || !warnedMessages.has(message)) {
								warnedMessages?.add(message);
								(options.error === false ? console.error : this.error.bind(this))(message);
							}
						}
						matched = true;
					}
				}
				return matched ? PROXY_ID : null;
			}
		};
	});
	if (traceEnabled) plugins.push({
		name: "impound:trace",
		enforce: "pre",
		resolveId(_id, importer, resolveOptions) {
			if (!importer && resolveOptions?.isEntry) entries.add(_id);
			return null;
		},
		async transform(code, id) {
			await init;
			try {
				const [imports] = parse(code, id);
				const importMap = /* @__PURE__ */ new Map();
				for (const imp of imports) if (imp.n) {
					const { line, column } = offsetToLineColumn(code, imp.s);
					importMap.set(imp.n, {
						line,
						column,
						statementStart: imp.ss,
						statementEnd: imp.se
					});
				}
				moduleGraph.set(id, {
					code,
					imports: importMap
				});
			} catch {}
		},
		buildEnd() {
			if (pendingViolations.length === 0) return;
			const warnedSets = /* @__PURE__ */ new Map();
			const errors = [];
			for (const violation of pendingViolations) {
				const { id, rawId, importer, relativeImporter, options, suggestions } = violation;
				const trace = buildTrace(importer, moduleGraph, resolvedImports, entries, maxTraceDepth);
				let snippet;
				const importerEntry = moduleGraph.get(importer);
				if (importerEntry) {
					const loc = importerEntry.imports.get(rawId);
					if (loc) snippet = {
						text: generateSnippet(importerEntry.code, loc.line, loc.column),
						line: loc.line,
						column: loc.column
					};
				}
				let message = violation.message;
				if (trace.length > 1) message += `\n\nTrace:\n${formatTrace(trace, globalOptions.cwd)}`;
				if (snippet) message += `\n\nCode:\n${snippet.text}`;
				if (suggestions?.length) message += `\n\nSuggestions:\n${suggestions.map((s) => `  - ${s}`).join("\n")}`;
				const violationInfo = {
					id,
					importer: relativeImporter,
					message,
					trace: trace.length > 1 ? trace : void 0,
					snippet
				};
				if (options.onViolation?.(violationInfo) === false) continue;
				let warnedMessages = warnedSets.get(options);
				if (options.warn !== "always") {
					if (!warnedMessages) {
						warnedMessages = /* @__PURE__ */ new Set();
						warnedSets.set(options, warnedMessages);
					}
					if (warnedMessages.has(message)) continue;
					warnedMessages.add(message);
				}
				if (options.error === false) console.error(message);
				else errors.push(message);
			}
			if (errors.length > 0) this.error(errors.join("\n\n---\n\n"));
		}
	});
	return plugins;
});

//#endregion
export { ImpoundPlugin };
